### ISSUE TITLE
netlink: towards using the network poller in 1.12

### ISSUE THREAD

**Issue by acln0** [mdlayher/netlink#119]
Hello.

[This CL](https://go-review.googlesource.com/c/go/+/155517) has landed, and will be included in Go 1.12. I think this opens an avenue for package netlink to make use of the poller.

I've investigated the code, and I believe the change can be made in a backwards-compatible manner, such that package netlink keeps working gracefully for < 1.12 users, and includes the new features for 1.12 and later.

In terms of changes to the public API of the package, no backwards incompatible changes would be made, but the set of patches would add `SetDeadline`, `SetReadDeadline` and `SetWriteDeadline` methods to `Conn`.

In terms of the internals of the package, making this change requires a pretty major refactor of the bits in `conn_linux.go`: `sysSocket` needs to change from using an integer file descriptor to an `*os.File`, and operations need to use `syscall.RawConn` when making system calls. A compatibility shim for pre-1.12 would also have to be written.

I am willing to commit to doing this work, provided that @mdlayher is open to the idea. I filed this bug in order to gauge interest, ask for a green light, and track progress.

Thanks,
Andrei

**Comment by mdlayher** [comment]
I'm totally on board with these changes, and honestly, I have no qualms making an immediate move to 1.12 and removing compatibility for any prior versions. Anybody who is concerned should already have the package vendored in their application.

Thanks again for doing this!

**Comment by acln0** [comment]
I have implemented most of the things described in the initial issue in this commit: https://github.com/acln0/netlink/commit/2a82be35c5523408593ddb98ff5b1a6a97b43d78

The patch is not ready for review just yet, but I wanted to write a few lines about it, as a progress report, and to discuss some more specific details. The good news is that everything seems to be working as expected. The patch is still lacking some tests, but most of the changes I had in mind are there, so it's pretty close to what I envision its final form to be. 

As for the actual topic of discussion: for now, I have kept the old behavior of the library for Go < 1.12, and introduced the new behavior for Go 1.12+. In my tree, both code paths use `*os.File`, but pre-1.12 code uses a blocking descriptor (and blocking calls), like it always has. I added the new deadline methods to Conn for all code, but they don't work on Go < 1.12.

If anyone wants to play with the code, note that it uses the "tip" build tag for now, because there is no Go 1.12 yet. This will change when 1.12 actually lands.

The build tags and file names I added are pretty ugly, but they are necessary if we want the code to keep building on < 1.12. I don't see a cleaner way to achieve backwards compatibility. @mdlayher, if these things are not too gross for you, I would propose that the package keep compatibility, at least for a while. Requiring a fresh new Go version seems a little harsh to me. I can commit to maintaining the compatibility code until we decide to pull the plug. I think the compatibility code is not a big deal, it's just a little ugly.

I'm going to send a PR when 1.12 lands, so we can test it with the actual released version of Go, and not from tip (which I find a little cumbersome). If all goes well, after this is reviewed, tested and merged, perhaps we should leave a note in https://github.com/golang/go/issues/15021 that the new 1.12 features are indeed good enough to solve all the problems.
